File: ios/TidalWidget/Shared/SharedState.swift
import Foundation
import WidgetKit

/// The Swift half of the shared-cache contract. Must stay in step with
/// lib/models/widget_payload.dart — Flutter is the only writer of `tidal_state`,
/// the widget is the only writer of `tidal_pending_open`.
///
/// The widget never talks to Firestore: it has no auth and no SDK. Everything it
/// knows arrives through this blob.
enum LetterState: String {
    case empty, waiting, open, faded
}

struct TidalState {
    var state: LetterState = .empty
    var messageId: String?
    /// Present whenever a letter exists, INCLUDING while waiting. Rendering is
    /// gated on `state`, never on the presence of this — that is what lets the
    /// cat wake instantly on tap without a network round trip.
    var text: String?
    var openedAt: Date?
    var expiresAt: Date?
    var partnerName: String?
    var partnerCatId: String?
    var idleLine: String?

    /// Client-side expiry. The cached blob can outlive its own window, so never
    /// trust `state` alone.
    func effectiveState(now: Date = Date()) -> LetterState {
        if state == .open, let e = expiresAt, now >= e { return .faded }
        return state
    }
}

enum SharedStore {
    static let appGroup = "group.com.mypeblo.tidal"
    static let stateKey = "tidal_state"
    static let pendingOpenKey = "tidal_pending_open"
    static let widgetKind = "TidalWidget"

    private static var defaults: UserDefaults? { UserDefaults(suiteName: appGroup) }

    private static func date(_ v: Any?) -> Date? {
        guard let ms = v as? Double, ms > 0 else { return nil }
        return Date(timeIntervalSince1970: ms / 1000)
    }

    static func load() -> TidalState {
        guard let raw = defaults?.string(forKey: stateKey),
              let data = raw.data(using: .utf8),
              let j = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return TidalState() }

        return TidalState(
            state: LetterState(rawValue: j["state"] as? String ?? "") ?? .empty,
            messageId: j["messageId"] as? String,
            text: j["text"] as? String,
            openedAt: date(j["openedAtMs"]),
            expiresAt: date(j["expiresAtMs"]),
            partnerName: j["partnerName"] as? String,
            partnerCatId: j["partnerCatId"] as? String,
            idleLine: j["idleLine"] as? String
        )
    }

    /// Local acknowledgement of a tap: the letter will be read in the app, so
    /// the widget's cat goes straight back to sleep, and the breadcrumb tells
    /// Flutter which letter to open and show.
    static func markTappedLocally() {
        guard let d = defaults,
              let raw = d.string(forKey: stateKey),
              let data = raw.data(using: .utf8),
              var j = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              (j["state"] as? String) == LetterState.waiting.rawValue,
              let messageId = j["messageId"] as? String
        else { return }

        let nowMs = Date().timeIntervalSince1970 * 1000
        j["state"] = LetterState.empty.rawValue

        if let out = try? JSONSerialization.data(withJSONObject: j),
           let s = String(data: out, encoding: .utf8) {
            d.set(s, forKey: stateKey)
        }

        // Written even if the state write above failed — losing the breadcrumb
        // means the open never reaches Firestore and the partner never sees it.
        if let crumb = try? JSONSerialization.data(withJSONObject: [
            "messageId": messageId,
            "tappedAtMs": nowMs,
        ]), let s = String(data: crumb, encoding: .utf8) {
            d.set(s, forKey: pendingOpenKey)
        }
    }

    static func reload() {
        WidgetCenter.shared.reloadTimelines(ofKind: widgetKind)
    }
}

File: android/app/src/main/kotlin/com/mypeblo/tidal/widget/SharedState.kt
package com.mypeblo.tidal.widget

import android.content.Context
import es.antonborri.home_widget.HomeWidgetPlugin
import org.json.JSONObject

/**
 * Android half of the shared-cache contract. Mirrors
 * ios/TidalWidget/Shared/SharedState.swift and lib/models/widget_payload.dart —
 * nothing links the three, so tools/check_payload_contract.py diffs them.
 *
 * Flutter is the only writer of `tidal_state`; the widget is the only writer of
 * `tidal_pending_open`.
 */
enum class LetterState { EMPTY, WAITING, OPEN, FADED;

    companion object {
        fun from(raw: String?) = when (raw) {
            "waiting" -> WAITING
            "open" -> OPEN
            "faded" -> FADED
            else -> EMPTY
        }
    }
}

data class TidalState(
    val state: LetterState = LetterState.EMPTY,
    val messageId: String? = null,
    /** Present even while waiting; rendering is gated on [state], never on this. */
    val text: String? = null,
    val expiresAtMs: Long = 0,
    val partnerName: String? = null,
    val partnerCatId: String? = null,
    val idleLine: String? = null,
) {
    /**
     * A cached blob can outlive its own window — the widget may not be updated
     * for hours — so expiry is always recomputed rather than trusted.
     */
    fun effectiveState(now: Long = System.currentTimeMillis()): LetterState =
        if (state == LetterState.OPEN && expiresAtMs in 1..now) LetterState.EMPTY else state
}

object SharedState {
    const val STATE_KEY = "tidal_state"
    const val PENDING_OPEN_KEY = "tidal_pending_open"

    private fun prefs(context: Context) = HomeWidgetPlugin.getData(context)

    fun load(context: Context): TidalState {
        val raw = prefs(context).getString(STATE_KEY, null) ?: return TidalState()
        return try {
            val j = JSONObject(raw)
            TidalState(
                state = LetterState.from(j.optString("state", "empty")),
                messageId = j.optStringOrNull("messageId"),
                text = j.optStringOrNull("text"),
                expiresAtMs = j.optLong("expiresAtMs", 0),
                partnerName = j.optStringOrNull("partnerName"),
                partnerCatId = j.optStringOrNull("partnerCatId"),
                idleLine = j.optStringOrNull("idleLine"),
            )
        } catch (_: Exception) {
            // A corrupt blob renders the quiet state rather than crashing the
            // launcher, which is what an uncaught exception here would do.
            TidalState()
        }
    }

    /**
     * Reveal in place: flip the cached state to `open` so the letter shows on
     * the widget immediately, give it a short reading window, and leave a
     * breadcrumb so Flutter flushes the real open to Firestore next time the
     * app runs. No app launch — reading happens right here.
     */
    fun markOpenedLocally(context: Context, readingWindowMs: Long) {
        val p = prefs(context)
        val raw = p.getString(STATE_KEY, null) ?: return
        try {
            val j = JSONObject(raw)
            if (j.optString("state") != "waiting") return
            val messageId = j.optStringOrNull("messageId") ?: return

            val now = System.currentTimeMillis()
            j.put("state", "open")
            j.put("openedAtMs", now)
            j.put("expiresAtMs", now + readingWindowMs)

            p.edit()
                .putString(STATE_KEY, j.toString())
                .putString(
                    PENDING_OPEN_KEY,
                    JSONObject()
                        .put("messageId", messageId)
                        .put("tappedAtMs", now)
                        .toString(),
                )
                .apply()
        } catch (_: Exception) {
            // Leave the cache untouched; the app reconciles from Firestore.
        }
    }

    private fun JSONObject.optStringOrNull(key: String): String? =
        if (isNull(key)) null else optString(key, "").ifEmpty { null }
}
